File: assignment3/Client/src/EncoderDecoder.cpp
//
// Created by david on 1/15/17.
//

#include <cstring>
#include "../include/EncoderDecoder.h"

//ACK , DATA , ERROR , BCAST
Packet *EncoderDecoder::decodeNextByte(char nextByte) {
    count++;
    bytes.push_back(nextByte);  //THIS IS A VECTOR
    Packet *p;
    if (count > 2) {

        opCode = byteToShort(bytes, 0, 1);
        if (opCode == 3) {
            if (count == 4) {
                size = byteToShort(bytes, 2, 3);
            }
            if (count == 6) {
                blockNum = byteToShort(bytes, 4, 5);
            }
            if (count == size + 6) {
                vector<char> newVec(bytes.begin() + 6, bytes.end());
                p = new DATA(size, blockNum, newVec);
                reset();
                return p;
            }
        } else if (opCode == 4) {
            short blockNum = 0;
            if (count == 4) {
                blockNum = byteToShort(bytes, 2, 3);
                p = new ACK(blockNum);
                reset();
                return p;
            }
        } else if (opCode == 5) {
            short errCode = 0;
            if (count == 4)
                errCode = byteToShort(bytes, 2, 3);
            if (nextByte == '\0' && count > 4) {
                std::string str(bytes.begin() + 4, bytes.end() - 1);
                p = new ERROR(errCode, str);
                reset();
                return p;
            }
        } else if(opCode == 9) {
            char flag = 0;
            if (count == 3)
                flag = nextByte;


            if (nextByte == '\0' && count > 3) {
                std::string str(bytes.begin() + 3, bytes.end() - 1);
                p = new BCAST(flag, str);
                reset();
                return p;
            }

        }
    }
    return nullptr;
}

char *EncoderDecoder::encode(Packet &pack, int& l) {
    opCode = pack.opCode;
    char *code = shortToBytes(opCode);
    char* c;
    if (opCode == 7 || opCode == 8 || opCode == 1 || opCode == 2) {
        vector<char> data;
        if (opCode == 2) {
            WRQ &p = dynamic_cast<WRQ &>(pack);
            str = p.getFileName();
        } else if (opCode == 1) {
            RRQ &p = dynamic_cast<RRQ &>(pack);
            str = p.getFileName();
        } else if (opCode == 7) {
            LOGRQ &p = dynamic_cast<LOGRQ &>(pack);
            str = p.getUserName();
        } else if (opCode == 8){
            DELRQ &p = dynamic_cast<DELRQ &>(pack);
            str = p.getFileName();
        }
        vector<char> name(str.begin(), str.end());
        vector<char> res(name.size() + 3);
        res[0] = code[0];
        res[1] = code[1];
        int length = res.size() - 1;
        for (int i = 2; i < length; i++)
            res[i] = name[i - 2];
        l = res.size();
        res[res.size() - 1] = '\0';
        c = new char[res.size()];
        std::copy(res.begin(), res.end(), c);
        reset();
        return c;
    } else if (opCode == 4) {
        ACK &p = dynamic_cast<ACK &>(pack);
        short t = p.getBlockNum();
        char *block = shortToBytes(t);
        c = new char[4];
        c[0] = code[0];
        c[1] = code[1];
        c[2] = block[0];
        c[3] = block[1];
        l = 4;
        reset();
        return c;
    } else if (opCode == 3) {
        DATA &p = dynamic_cast<DATA &>(pack);
        char *size = shortToBytes(p.getSize());
        char *blockNum = shortToBytes(p.getBlockNum());
        vector<char> data = p.getData();
        vector<char> res(data.size() + 6);
        res[0] = code[0];
        res[1] = code[1];
        res[2] = size[0];
        res[3] = size[1];
        res[4] = blockNum[0];
        res[5] = blockNum[1];
        short length = res.size();
        for (int i = 6; i < length; i++) {
            res[i] = data[i - 6];
        }
        l = res.size();
        c = new char[res.size()];
        copy(res.begin(), res.end(), c);
        reset();
        return c;
    } else if (opCode == 6 || opCode == 10) {
        c = code;
        l = 2;
        reset();
        return c;
    } else {
        ERROR &p = dynamic_cast<ERROR &>(pack);
        char *errC = shortToBytes(p.getErrCode());
        str = p.getErrMsg();
        vector<char> name(str.begin(), str.end());
        vector<char> res(name.size() + 5);
        res[0] = code[0];
        res[1] = code[1];
        res[2] = errC[0];
        res[3] = errC[1];
        short l = res.size() - 1;
        for (int i = 2; i < l; i++) {
            res[i] = name[i - 4];
        }
        res[res.size() - 1] = '\0';
        c = new char[res.size()];
        copy(res.begin(), res.end(), c);
        l = res.size();
        reset();
        return c;
    }
}

short EncoderDecoder::byteToShort(vector<char> bytesArr, int a, int b) {
    short res = (short) ((bytesArr[a] & 0xff) << 8);
    res += (short) (bytesArr[b] & 0xff);
    return res;
}

char * EncoderDecoder::shortToBytes(short num) {
    char *bytesArr = new char[2];
    bytesArr[0] = ((num >> 8) & 0xFF);
    bytesArr[1] = (num & 0xFF);
    return bytesArr;
}

void EncoderDecoder::reset() {
    bytes.clear();
    count = 0;
    opCode = 0;
    str = "";
    size = 0;
    blockNum = 0;
    errCode = 0;
}

File: assignment3/Server/src/main/java/bgu/spl171/net/impl/TFTPreactor/EncoderDecoder.java
package bgu.spl171.net.impl.TFTPreactor;

import bgu.spl171.net.api.MessageEncoderDecoder;
import bgu.spl171.net.impl.packet.*;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;

/**
 * Created by david on 1/13/17.
 */
public class EncoderDecoder implements MessageEncoderDecoder<Packet> {

    private byte[] bytes = new byte[518]; //start with 1k
    private int len;
    private int count;
    private short opCode;
    private String str;
    private short size;
    private short blockNum;



    private int counter = 0;

    public EncoderDecoder() {
        len = 0;
        count = 0;
        opCode = 0;
        str = "";
        size = 0;
        blockNum = 0;
    }

    public Packet decodeNextByte(byte nextByte) {
        count++;
        pushByte(nextByte);
        if (count >= 2) {
            opCode = byteToShort(bytes, 0, 1);
            if (opCode == 1 || opCode == 2 || opCode == 7 || opCode == 8) {
                str = byteToString(nextByte, 2);
                if (str != null) {
                    Packet tmp;
                    if (opCode == 1)
                        tmp = new RRQ(str);
                    else if (opCode == 2)
                        tmp = new WRQ(str);
                    else if (opCode == 7)
                        tmp = new LOGRQ(str);
                    else
                        tmp = new DELRQ(str);
                    reset();
                    return tmp;
                }
            } else if (opCode == 3) {
                if (count == 4) {
                    size = byteToShort(bytes, 2, 3);
                }
                if (count == 6) {
                    blockNum = byteToShort(bytes, 4, 5);
                }
                if (count == size + 6) {
                    byte[] data = Arrays.copyOfRange(bytes, 6, size + 6);
                    DATA tmp = new DATA(size, blockNum, data);
                    reset();
                    return tmp;
                }
            } else if (opCode == 4) {
                if (count == 4) {
                    blockNum = byteToShort(bytes, 2, 3);
                    ACK tmp = new ACK(blockNum);
                    reset();
                    return tmp;
                }
            } else if (opCode == 5) {
                short errCode = 0;
                if (count == 4)
                    errCode = byteToShort(bytes, 2, 3);
                if (count > 4)
                    str = byteToString(nextByte, 4);
                // error massage generares automaticly by the errcode.
                ERROR tmp = new ERROR(errCode);
                reset();
                return tmp;
            } else if (opCode == 6) {
                DIRQ tmp = new DIRQ();
                reset();
                return tmp;
            } else if (opCode == 10) {
                DISC tmp = new DISC();
                reset();
                return tmp;
            } else {
                ERROR tmp = new ERROR((short) 4);
                return tmp;
            }
        }
        return null;
    }

    public byte[] encode(Packet message) {
        short opCode = message.getOpcode();
        byte[] code = shortToBytes(opCode);
        byte[] res;
        if (opCode == 3) {
            byte[] size = shortToBytes(((DATA) message).getPacketSize());
            byte[] blockNum = shortToBytes(((DATA) message).getBlockNum());
            byte[] data = ((DATA) message).getData();
            res = merge(code, size);
            res = merge(res, blockNum);
            res = merge(res, data);
            counter += data.length;
        } else if (opCode == 4) {
            byte[] blockNum = shortToBytes(((ACK) message).getBlockNum());
            res = merge(code, blockNum);
        } else if (opCode == 5) {
            byte[] str = stringToByte(((ERROR) message).getErrMsg());
            byte[] lastByte = {0};
            byte[] errCode = shortToBytes(((ERROR) message).getErrorCode());
            res = merge(code, errCode);
            res = merge(res, str);
            res = merge(res, lastByte);
        } else {
            byte[] str = stringToByte(((BCAST) message).getFilename());
            byte[] lastByte = {0};
            byte[] flag = {((BCAST) message).getFlag()};
            res = merge(code, flag);
            res = merge(res, str);
            res = merge(res, lastByte);
        }
        return res;
    }

    public static byte[] merge(byte[] first, byte[] second) {
        byte[] res = new byte[first.length + second.length];
        for (int i = 0; i < res.length; i++) {
            if (first.length > i)
                res[i] = first[i];
            else
                res[i] = second[i - first.length];
        }
        return res;
    }

    public short byteToShort(byte[] byteArr, int a, int b) {
        short res = (short) ((byteArr[a] & 0xff) << 8);
        res += (short) (byteArr[b] & 0xff);
        return res;
    }

    public static byte[] shortToBytes(short message) {
        byte[] bytesArr = new byte[2];
        bytesArr[0] = (byte) ((message >> 8) & 0xFF);
        bytesArr[1] = (byte) (message & 0xFF);
        return bytesArr;
    }

    public String byteToString(byte nextByte, int start) {
        //notice that the top 128 ascii characters have the same representation as their utf-8 counterparts
        //this allow us to do the following comparison
        if (nextByte == '\0') {
            //notice that we explicitly requesting that the string will be decoded from UTF-8
            //this is not actually required as it is the default encoding in java.
            String result = new String(bytes, start, len - 3, StandardCharsets.UTF_8);
            len = 0;
            return result;
        }
        return null; //not a line yet
    }

    public static byte[] stringToByte(String message) {
        return (message + "\n").getBytes(); //uses utf8 by default
    }


    private void pushByte(byte nextByte) {
        if (len >= bytes.length)
            bytes = Arrays.copyOf(bytes, len * 2);
        bytes[len++] = nextByte;
    }

    private void reset() {
        bytes = new byte[518]; //start with 1k
        len = 0;
        count = 0;
        opCode = 0;
        str = null;
        size = 0;
        blockNum = 0;
    }
}
